File: web/js/views/health.js
/**
 * Health view.
 *
 * Fetches GET /api/health (via apiFetch, so the backend resolves THIS account's own archive - see api/routes/health.py's module docstring)
 * and shows the API's own liveness report:
 * whether the caller's archive is attached and readable, whether a Telethon session exists, and the current archived message count.
 * See api/routes/health.py for exactly what each check does and doesn't cover
 * (notably: it does NOT confirm the userbot is currently connected to Telegram — that needs IPC, a Phase 3 addition per that file's own docstring).
 *
 * No polling — this is a manually-refreshed diagnostic view, not a live dashboard.
 * Deliberately simple: one fetch, one refresh button.
 * Adding auto-refresh is a small change later if it turns out to be wanted, not something to build speculatively now.
 *
 * Lazy-initialized by app.js on first tab open, same pattern as the other non-landing views.
 */

import { t } from "../i18n.js";
import { apiFetch } from "../lib/auth.js";
import { describeError } from "../lib/errors.js";

const healthViewState = {
  initialized: false,
};

/**
 * Render the health check results as an HTML string.
 * @param {object} data - a HealthOut record from the API.
 * @returns {string}
 */
function renderHealthReport(data) {
  const statusLabel =
    data.status === "ok" ? t("health.statusOk") : t("health.statusDegraded");

  const checkRow = (label, passed) => `
    <li class="health-check">
      <span class="health-check__mark" aria-hidden="true">${passed ? "✓" : "✗"}</span>
      <span>${label}</span>
    </li>
  `;

  // is_instance_owner: there is exactly one physical Telethon session per running instance, and it belongs to one account (see HealthOut's own docstring) -
  // showing a plain ✗ to every OTHER account would look like a persistent, personal problem rather than "not applicable to you",
  // so those accounts get an explanatory line instead of a checklist row here.
  const sessionRow = data.is_instance_owner
    ? checkRow(t("health.sessionExists"), data.session_exists)
    : `
    <li class="health-check health-check--info">
      <span class="health-check__mark" aria-hidden="true">·</span>
      <span>${t("health.sessionNotApplicable")}</span>
    </li>
  `;

  // archive_status distinguishes WHY the database check didn't pass (unattached vs. genuinely unavailable)
  // instead of collapsing both into one flat "not readable" line - see HealthOut's own docstring.
  // "ok" is the only case with an actual message count to show.
  const archiveMessage =
    data.archive_status === "unattached"
      ? `<p class="health-archive-message">${t("health.unattached")}</p>`
      : data.archive_status === "unavailable"
        ? `<p class="health-archive-message">${t("health.unavailable")}</p>`
        : `<p class="health-message-count">${t("health.messageCount")}: ${data.db_message_count}</p>`;

  return `
    <div class="health-status">
      <span class="info-badge">${statusLabel}</span>
    </div>
    <ul class="health-check-list">
      ${checkRow(t("health.dbReadable"), data.db_readable)}
      ${sessionRow}
    </ul>
    ${archiveMessage}
    <button id="health-refresh" class="health-refresh-btn" type="button">${t("health.refresh")}</button>
  `;
}

/** Fetch and render the current health report. */
async function loadHealth(root) {
  root.innerHTML = `<div class="empty-state">${t("common.loading")}</div>`;

  let data;
  try {
    // apiFetch (not plain fetch) - GET /api/health requires a token like every other route now that each account has its own,
    // potentially-nonexistent archive (see api/routes/health.py's module docstring):
    // there's no database to check without first knowing which caller is asking.
    const res = await apiFetch("/api/health");
    // Not using res.ok for the normal case — health.py always returns 200, even when status is "degraded"
    // (that's the point: the body carries the real state, not the HTTP status — see its docstring).
    // A non-200 here means something is more seriously wrong, e.g. an expired session that apiFetch couldn't silently refresh
    // (rare on this view - see apiFetch's own docstring for when it redirects instead of returning here).
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      root.innerHTML = `<div class="empty-state">${describeError(body.detail)}</div>`;
      return;
    }
    data = await res.json();
  } catch {
    root.innerHTML = `<div class="empty-state">${t("common.error")}</div>`;
    return;
  }

  root.innerHTML = renderHealthReport(data);
  root
    .querySelector("#health-refresh")
    .addEventListener("click", () => loadHealth(root));
}

/** Entry point called by app.js the first time the Health tab is opened. */
function initHealthView() {
  if (healthViewState.initialized) return;
  healthViewState.initialized = true;

  const root = document.getElementById("health-root");
  if (root) loadHealth(root);
}

export { initHealthView };

// Re-render on language change.
// Re-fetches (rather than caching like the other views) since this is a live diagnostic — the whole point of a health check is that it reflects the current moment,
// not the last time the tab happened to be opened.
document.addEventListener("televault:langchange", () => {
  if (!healthViewState.initialized) return;
  const root = document.getElementById("health-root");
  if (root) loadHealth(root);
});

File: api/routes/health.py
"""
GET /api/health — liveness check for the API server and the caller's own archive.

Used by:
    - The web UI to show a status indicator (per-account, like /chats, /messages, /deleted, /stats)
    - Future: heartbeat from the userbot process (Phase 3)

Requires a valid access token, same as every other archive-facing route (see api/dependencies.py's get_current_user()).
This is NOT how the route started: it was originally open (no token required) back when there was exactly one archive per running instance and "is the process alive"
was a meaningful question on its own.
Now that every account has its own, potentially-nonexistent archive_db_ref, "is THE database healthy" isn't a question with a single answer any more -
there's no database to check without first knowing which caller is asking.
A monitoring script that wants a pure liveness probe with no archive semantics should hit something that doesn't need a token
(e.g. the process simply being up and answering HTTP at all) rather than this route - see CHANGELOG for whether/when a dedicated one gets added.

What this checks vs. what it doesn't:
    ✓ The caller's own archive database (resolved via control_db.users.archive_db_ref, same lookup get_archive_connection() uses for /chats, /messages, /deleted, /stats)
    is readable and returns rows
    ✓ Telethon .session file exists on disk (instance-level - see session_exists' own note below)
    ✗ Whether the userbot is currently connected to Telegram (that requires IPC — a Phase 3 addition, see CHANGELOG)
"""

from pathlib import Path

from fastapi import APIRouter, Depends
from sqlalchemy.engine import Connection
from sqlalchemy.exc import OperationalError

import control_db
import db
from api.dependencies import get_control_db, get_current_user
from api.schemas import HealthOut
from config import settings
from utils.security import DecodedAccessToken

router = APIRouter(tags=["health"])


@router.get("/health", response_model=HealthOut, summary="API and caller's archive liveness")
def health_check(
    current: DecodedAccessToken = Depends(get_current_user),
    control_conn: Connection = Depends(get_control_db),
) -> HealthOut:
    """
    Return liveness status for the API and the caller's own archive database.

    `archive_status` is one of:
        - "ok"          the caller's archive was reached and read successfully (db_message_count is accurate, including 0 - an attached-but-empty archive is NOT "unavailable")
        - "unattached"  this account has no archive_db_ref yet - nothing to check
        - "unavailable" an archive_db_ref exists but couldn't actually be reached right now

    `status` stays 'ok'/'degraded' for backwards compatibility with the existing UI badge -
    'ok' requires both archive_status == "ok" and the Telethon session file being present.
    A 200 response is always returned once the caller is authenticated - the body carries the real state, not the HTTP status,
    so partial-health states (e.g. "unattached") render as a normal page state in the UI rather than an error.
    (An invalid/expired/missing token itself still 401s, via get_current_user() above, same as every other authenticated route.)
    """
    message_count: int | None = None

    # Same lookup get_archive_connection() does for /chats, /messages, /deleted, /stats, but returning a status instead of raising 409/503:
    # this is a dashboard view, not a data endpoint, so "not set up yet" is a state to display, not an error.
    user = control_db.queries.get_user_by_id(control_conn, current["user_id"])
    archive_ref = user["archive_db_ref"] if user else None

    if archive_ref is None:
        archive_status = "unattached"
    else:
        try:
            with db.get_tenant_readonly_connection(archive_ref) as tenant_conn:
                message_count = db.get_message_count(tenant_conn)
            archive_status = "ok"
        except OperationalError:
            archive_status = "unavailable"

    # Telethon session file: instance-level, not per-user - there is one physical userbot process per instance today
    # (see require_instance_owner()'s docstring in api/dependencies.py).
    # Presence means the userbot has authenticated at least once;
    # absence most likely means first-run setup hasn't completed.
    # Shown to every caller regardless of archive_status - it describes this instance's userbot process, not any individual account.
    session_path = Path(settings.session_name).with_suffix(".session")
    session_exists = session_path.exists()

    db_readable = archive_status == "ok"
    status = "ok" if (db_readable and session_exists) else "degraded"

    return HealthOut(
        status=status,
        archive_status=archive_status,
        db_readable=db_readable,
        session_exists=session_exists,
        db_message_count=message_count,
    )